Youtube 360° video goes black

jernoble opened the pull request:
#### b2a7036dc8ddf82a440919da370702bd4c402854
<pre>
Youtube 360° video goes black
<a href="https://bugs.webkit.org/show_bug.cgi?id=259298">https://bugs.webkit.org/show_bug.cgi?id=259298</a>
rdar://112239644

Reviewed by Eric Carlson.

YouTube, when pre-rolling an advertisement before a 360 video, will start with
an element that&apos;s displayed normally, but will hide the video with display:none
when switching to canvas-based rendering for 360 videos. But at this point, the
MediaPlayerPrivate&apos;s layer has already been created and sized, and nothing will
force it to switch back to a decompression session (which is the only way for
frames to be extracted during playback).

When a video element&apos;s renderer is detached, queue a task to notify the media
player that the accelerated rendering state has changed. Similarly, do so when
changing the fullscreen mode. In MediaPlayerPrivateMediaSourceAVFObjC, use this
signal to tear down the AVSBDL when it is no longer needed.

* LayoutTests/media/media-source/media-source-paint-after-display-none-expected.txt: Added.
* LayoutTests/media/media-source/media-source-paint-after-display-none.html: Added.
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::didDetachRenderers):
(WebCore::HTMLMediaElement::setFullscreenMode):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::shouldEnsureLayer const):

Canonical link: <a href="https://commits.webkit.org/266162@main">https://commits.webkit.org/266162@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/09366e2988adaaadd58d222ca98c5bd883b6375a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13023 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13341 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13672 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14761 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12416 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13091 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15847 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13367 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15108 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13190 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13882 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11008 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15214 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11170 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11763 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18827 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12245 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11929 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15140 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12409 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10293 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11677 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11629 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3199 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15995 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12252 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on LayoutTests/media/media-source/media-source-paint-after-display-none.html:
```diff
@@ -0,0 +1,94 @@
+        video.play();
+        await waitFor(video, 'playing');
+
+        await sleepFor(200);
```
A hard coded sleep may cause problems on a heavily loaded bot. Maybe wait until the image data has changed, with a timeout?

eric-carlson reviewed (approved):
r=me once the bots are happy

eric-carlson commented on LayoutTests/media/media-source/media-source-paint-after-display-none.html:
```diff
@@ -0,0 +1,99 @@
+    async function runTest() {
+        findMediaElement();
+
+        var response = await fetch('content/test-red-3s-480x360.mp4');
```
Nit: var -> let

eric-carlson commented on LayoutTests/media/media-source/media-source-paint-after-display-none.html:
```diff
@@ -0,0 +1,99 @@
+            waitFor(sourceBuffer, 'update', true),
+            waitFor(video, 'canplay')
+        ]);
+
```
Nit: extra blank line

eric-carlson commented on LayoutTests/media/media-source/media-source-paint-after-display-none.html:
```diff
@@ -0,0 +1,99 @@
+    var source;
+    var sourceBuffer;
+
+    function loaderPromise(loader) {
+        return new Promise((resolve, reject) => {
+            loader.onload = resolve;
+            loader.onerror = reject;
+        });
+    }
```
It looks like this isn't used

jernoble committed:
Youtube 360° video goes black
https://bugs.webkit.org/show_bug.cgi?id=259298
rdar://112239644

Reviewed by Eric Carlson.

YouTube, when pre-rolling an advertisement before a 360 video, will start with
an element that's displayed normally, but will hide the video with display:none
when switching to canvas-based rendering for 360 videos. But at this point, the
MediaPlayerPrivate's layer has already been created and sized, and nothing will
force it to switch back to a decompression session (which is the only way for
frames to be extracted during playback).

When a video element's renderer is detached, queue a task to notify the media
player that the accelerated rendering state has changed. Similarly, do so when
changing the fullscreen mode. In MediaPlayerPrivateMediaSourceAVFObjC, use this
signal to tear down the AVSBDL when it is no longer needed.

* LayoutTests/media/media-source/media-source-paint-after-display-none-expected.txt: Added.
* LayoutTests/media/media-source/media-source-paint-after-display-none.html: Added.
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::didDetachRenderers):
(WebCore::HTMLMediaElement::setFullscreenMode):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::shouldEnsureLayer const):

Canonical link: https://commits.webkit.org/266162@main
